cram/cram_decode.c

```diff
@@ -2442,14 +2442,34 @@ int cram_decode_slice(cram_fd *fd, cram_container *c, cram_slice *s,
 		    && cr->ref_id >= 0
 		    && cr->ref_id != last_ref_id) {
 		    if (!fd->no_ref) {
-			if (!refs[cr->ref_id])
-			    refs[cr->ref_id] = cram_get_ref(fd, cr->ref_id,
-							    1, 0);
-			s->ref = refs[cr->ref_id];
+			pthread_mutex_lock(&fd->range_lock);
+			if (fd->range.refid < 0 || cr->ref_id == fd->range.refid) {
+			    pthread_mutex_unlock(&fd->range_lock);
+			    if (!refs[cr->ref_id])
+				refs[cr->ref_id] = cram_get_ref(fd, cr->ref_id,
+								1, 0);
+			    if (!(s->ref = refs[cr->ref_id])) {
+				pthread_mutex_unlock(&fd->range_lock);
+				return -1;
+			    }
+			} else {
+			    pthread_mutex_unlock(&fd->range_lock);
+			    // For multi-ref containers, we don't need to fetch all
+			    // refs if we're only querying one.
+			    s->ref = NULL;
+			}
 
+			pthread_mutex_lock(&fd->range_lock);
```

Comment:
Some functions like cram_get_ref can be slow and also contain various mutexes.  I want to keep the locking as small as possible, but it can be refactored to be cleaner.